src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -216,6 +260,102 @@ async def bsp_scalac_options_request(request: ScalacOptionsParams) -> ScalacOpti
     return ScalacOptionsResult(items=tuple(result.item for result in results))
 
 
+# -----------------------------------------------------------------------------------------------
+# Dependency Modules
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class ScalaBSPDependencyModulesRequest(BSPDependencyModulesRequest):
+    field_set_type = ScalaMetadataFieldSet
+
+
+def get_entry_for_coord(
+    lockfile: CoursierResolvedLockfile, coord: Coordinate
+) -> CoursierLockfileEntry | None:
+    for entry in lockfile.entries:
+        if entry.coord == coord:
+            return entry
+    return None
+
+
+@rule
+async def scala_bsp_dependency_modules(
+    request: ScalaBSPDependencyModulesRequest,
+    build_root: BuildRoot,
+) -> BSPDependencyModulesResult:
+    coarsened_targets = await Get(
+        CoarsenedTargets, Addresses([fs.address for fs in request.field_sets])
+    )
+    resolve = await Get(CoursierResolveKey, CoarsenedTargets, coarsened_targets)
+    lockfile = await Get(CoursierResolvedLockfile, CoursierResolveKey, resolve)
+
+    # TODO: Can this use ClasspathEntryRequest?
+    transitive_targets = await Get(
+        TransitiveTargets,
+        TransitiveTargetsRequest(
+            roots=[
+                tgt.address
+                for coarsened_target in coarsened_targets
+                for tgt in coarsened_target.members
+            ]
+        ),
+    )
```

Comment:
`CoarsenedTargets` already contains the transitive deps (recursively structure shared): to get them, can do:
```
transitive_targets = [t for ct in coarsened_targets.closure() for t in ct.members]
```

Reply:
Nice! TIL. Will fix in a follow-on.